File: ios/Classes/RppgCommonPlugin.swift
import Flutter
import AVFoundation
import UIKit
import RPPGCommon
import RPPGCore

public class RppgCommonPlugin: NSObject, FlutterPlugin {

    let shared = Analysis.shared
    
    // MARK: Singleton Instance:
    /// Defines a shared instance of the RppgCommonPlugin class using the singleton pattern
    static let rppgPluginShared : RppgCommonPlugin = {
        let instace = RppgCommonPlugin()
        
        return instace
    }()
    
    /// Declare our eventSink, it will be initialized later
    private var eventSink: FlutterEventSink?
    
    override init() {
        super.init()
    }
        
    public static func register(with registrar: FlutterPluginRegistrar) {
        /// Method Channel
        let channel = FlutterMethodChannel(name: "rppg_common", binaryMessenger: registrar.messenger())
        registrar.addMethodCallDelegate(rppgPluginShared, channel: channel)
        
        /// Event Channel
        let eventChannel = FlutterEventChannel(name: "rppgCommon/analysisDataStream", binaryMessenger: registrar.messenger())
        eventChannel.setStreamHandler(rppgPluginShared)
                
        /// Platform View
        let factory = FLNativeViewFactory(messenger: registrar.messenger())
        registrar.register(factory, withId: "rppg_native_camera")
    
    }

    
    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "getState":
            result(shared.getState())
        case "askPermissions":
            result(shared.askPermissions())
        case "configure":
            if let arguments = call.arguments as? [String: Any] {
                if let fps = arguments["fps"] as? Int {
                    if let isFrontCamera = arguments["isFrontCamera"] as? Bool {
                        var userCameraPosition: RPPGVideoSessionSettings.CameraPosition = .front
                        if isFrontCamera == true {
                            userCameraPosition = .front
                        } else {
                            userCameraPosition = .back
                        }
                        shared.configure(CMTimeScale(fps), userCameraPosition)
                    }
                }
            }
        case "startVideo":
            shared.startVideo()
        case "startAnalysis":
            if let arguments = call.arguments as? [String: Any] {
                if let baseUrl = arguments["baseUrl"] as? String {
                    if let authToken = arguments["authToken"] as? String {
                        if let fps = arguments["fps"] as? String {
                            if let age = arguments["age"] as? String {
                                if let sex = arguments["sex"] as? String {
                                    if let height = arguments["height"] as? String {
                                        if let weight = arguments["weight"] as? String {
                                            shared.startAnalysis(baseUrl: baseUrl, authToken: authToken, fps: fps, age: age, sex: sex, height: height, weight: weight)
                                        }
                                    }
                                }
                            }
                        }
                    }
                }
            }
        case "stopAnalysis":
            shared.stopAnalysis()
        case "cleanMesh":
            shared.cleanMesh()
        case "meshColor":
            shared.meshColor()
        default:
            result(FlutterMethodNotImplemented)
        }
    }

}

//MARK: - FlutterStreamHandler

extension RppgCommonPlugin: FlutterStreamHandler {
   
    public func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        self.eventSink = events
        return nil
    }
    
    public func onCancel(withArguments arguments: Any?) -> FlutterError? {
        self.eventSink = nil
        return nil
    }
    
    /// Method to send dynamic data
    public func sendEvent(data: Any) {
        /// Check if the eventSink is available
        guard let eventSink = self.eventSink else {
            return
        }
        eventSink(data)
    }
    
    /// Method to stop the listener
    public func stopListening() {
        /// Check if the eventSink is available
        guard eventSink != nil else {
            return
        }
        /// Send a closing event
        ///
        /// Release resources if needed
        ///
        /// Set eventSink to nil to indicate that the listener has been stopped
        self.eventSink = nil
    }
}

File: android/src/main/kotlin/com/example/rppg_common/RppgCommonPlugin.kt
package com.example.rppg_common

import android.app.Activity
import com.example.rppg_common.utils.PermissionResult
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result


/** RppgCommonPlugin */
class RppgCommonPlugin: FlutterPlugin, MethodCallHandler, ActivityAware,
  EventChannel.StreamHandler {

  private var analysisInstance = Analysis.getInstance()

  /// Method channel
  private lateinit var channel : MethodChannel

  /// Event channel
  private var eventChannel: EventChannel? = null
  var eventSink: EventChannel.EventSink? = null


  /// Activity
  private lateinit var activity: Activity

  /// App Lifecycle
  private lateinit var lifecycle: Any

  override fun onDetachedFromActivity() {
    TODO("Not yet implemented")
  }

  override fun onReattachedToActivityForConfigChanges(binding: ActivityPluginBinding) {
    TODO("Not yet implemented")
  }

  override fun onAttachedToActivity(binding: ActivityPluginBinding) {
    activity = binding.activity;
    lifecycle = binding.lifecycle;
    if(this::activity.isInitialized) {
      analysisInstance.initialization(activity, lifecycle)
    }
  }

  override fun onDetachedFromActivityForConfigChanges() {

  }

  override fun onAttachedToEngine(flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
    // Method Channel
    channel = MethodChannel(flutterPluginBinding.binaryMessenger, "rppg_common")
    channel.setMethodCallHandler(this)

    // Event Channel
    eventChannel = EventChannel(flutterPluginBinding.binaryMessenger, "rppgCommon/analysisDataStream"); // timeHandlerEvent event name
    eventChannel?.setStreamHandler(this)

    // Platform View
    flutterPluginBinding
      .platformViewRegistry
      .registerViewFactory("rppg_native_camera", FLNativeViewFactory())

  }

  override fun onMethodCall(call: MethodCall, result: Result) {

    when (call.method) {
      "getState"->
        result.success(analysisInstance.getState())
      "askPermissions"->
        analysisInstance.setupPermission(activity,object :PermissionResult{
          override fun onPermissionCheck(isGranted: Boolean) {
            result.success(isGranted)
          }
        })
      "configure"-> {
        var fps = call.argument<Int>("fps")
        var isFrontCamera = call.argument<Boolean>("isFrontCamera")
        if (fps != null) {
          if (isFrontCamera != null) {
            analysisInstance.configure(fps = fps,isFrontCamera = isFrontCamera)
          }
        }
      }
      "startVideo"-> {
        analysisInstance.startVideo(activity)
      }
      "startAnalysis"-> {
        var baseUrl = call.argument<String>("baseUrl").toString()
        var authToken = call.argument<String>("authToken").toString()
        var fps = call.argument<String>("fps").toString()
        var age = call.argument<String>("age").toString()
        var sex = call.argument<String>("sex").toString()
        var height = call.argument<String>("height").toString()
        var weight = call.argument<String>("weight").toString()

        analysisInstance.startAnalysis(
          baseUrl,
          authToken,
          fps,
          age,
          sex,
          height,
          weight
          )
      }
      "stopAnalysis"-> {
        analysisInstance.stopAnalysis()
      }
      "cleanMesh"-> {
        analysisInstance.cleanMesh()
      }
      "meshColor"-> {
        analysisInstance.meshColor()
      }
      else ->
        result.notImplemented()
    }

  }

  override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
    channel.setMethodCallHandler(null)
  }


  /// Event Channel
  override fun onListen(arguments: Any?, eventSink: EventChannel.EventSink?) {
    this.eventSink = eventSink
    analysisInstance.setEventSink(eventSink)
  }

  override fun onCancel(arguments: Any?) {
    eventSink = null
    eventChannel = null
  }

  public fun sendEvent(data: Any) {
    // Check if the eventSink is available
    if(eventSink == null) {
      return
    }

    eventSink!!.success(data)
  }

  public fun cancelStream() {
    if(eventSink == null) {
      return
    }

    eventSink!!.endOfStream()
  }


}
